README_cmake.md

```diff
@@ -183,61 +454,141 @@ cases, you must declare that two Halide libraries share a runtime, like so:
+`my_gen` and a generator within called `my_filter`. Then you can pass a variable
+name to the `REGISTRATION` parameter of `add_halide_library` which will contain
+the name of a generated C++ source that should be linked to `Halide::RunGenMain`
+and `my_filter`.
+
+For example:
+
+```
+add_halide_library(my_filter FROM my_gen
+                   REGISTRATION filter_reg_cpp)
+add_executable(runner ${filter_reg_cpp})
+target_link_libraries(runner PRIVATE my_filter Halide::RunGenMain)
+```
 
-## Imported Targets
+Then you can run, debug, and benchmark your generator through the `runner`
+executable.
+
+## Halide package documentation
+
+### Components
+
+The Halide package script understands a handful of optional components when
+loading the package.
+
+First, if you plan to use the Halide Image IO library, you will want to include
+the `png` and `jpeg` components when loading Halide.
+
+Second, Halide releases can contain a variety of configurations: static, shared,
+debug, release, etc. CMake handles Debug/Release configurations automatically,
+but generally only allows one type of library to be loaded.
+
+The package understands two components, `static` and `shared`, that specify
+which type of library you would like to load. For example, if you want to make
+sure that you link against shared Halide, you can write:
+
+```
+find_package(Halide REQUIRED COMPONENTS shared)
+```
+
+If the shared libraries are not available, this will result in a failure.
+
+If no component is specified, then the `Halide_SHARED_LIBS` variable is checked.
+If it is defined and set to true, then the shared libraries will be loaded or
+the package loading will fail. Similarly, if it is defined and set to false, the
+static libraries will be loaded.
+
+If no component is specified and `Halide_SHARED_LIBS` is _not_ defined, then the
+`BUILD_SHARED_LIBS` variable will be inspected. If it is **not defined** or
+**defined and set to true**, then it will attempt to load the shared libs and
+fall back to the static libs if they are not available. Similarly, if
+`BUILD_SHARED_LIBS` is **defined and set to false**, then it will try the static
+libs first then fall back to the shared libs.
+
+### Variables
+
+Variables that control package loading:
+
+- `Halide_SHARED_LIBS` -- override `BUILD_SHARED_LIBS` when loading the Halide
+  package via `find_package`. Has no effect when using Halide via
+  `add_subdirectory` as a Git or FetchContent submodule.
+
+Variables set by the package:
+
+- `Halide_VERSION` -- The full version string of the loaded Halide package
+- `Halide_VERSION_MAJOR` -- The major version of the loaded Halide package
+- `Halide_VERSION_MINOR` -- The minor version of the loaded Halide package
+- `Halide_VERSION_PATCH` -- The patch version of the loaded Halide package
+- `Halide_VERSION_TWEAK` -- The tweak version of the loaded Halide package
+- `Halide_HOST_TARGET` -- The Halide target triple corresponding to "host" for
```

Comment:
I don't understand this entirely. Is this detected at the time you use the package, or at the time the package was built? (host could be different in the two cases, and the one at package creation time is not meaningful for use of Halide)

I ask because the other variables set are properties known at the time the Halide package was created.

Reply:
These are set when the package is created. The meaning of "host" is actually fixed when Halide is compiled, per our discussion on Halide/build on Gitter. Stamping that down at package creation time is the correct behavior. Detecting the host machine when the package is loaded is wrong for 32-bit Halide on 64-bit Windows

Reply:
I don't understand how that target triple could be useful then (due to the feature flags being potentially wrong). What is it used for?

Reply:
One issue is that just saying "host" doesn't guarantee that you aren't cross-compiling and you can get linker errors re: 32-bit/64-bit mismatch on Windows doing that.

But what's stored in `Halide_HOST_TARGET` is _just_ the triple and doesn't include feature flags. This means that when `host` wouldn't be cross-compiling, it can use "`host`" as the target that's passed to the generator, which lets it take full advantage of the host CPU.